### PR TITLE
[Supplier] Implement supplier service activation

### PR DESCRIPTION
**Pull request by red-0ne**
## Summary

This PR modifies the functionality so that suppliers staking for services mid-session will become active and be included in `Session.Suppliers` starting from the next session.

It adds the property `ServiceActivationHeight` which is a map of `serviceId` to `activationHeight` to the `Supplier` type and supports activation for both staking and restaking.

A remaining issue, is the immediate effect of removing a service when a Supplier restakes.

*~550 LOC are protobuf generated code*

## Issue

When a `Supplier` stakes or restakes mid-session, the session endpoint may include that supplier before the next session starts. This can lead to inconsistencies between session queries and clients that fetch `Session` data only when a new session begins.

- #695 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [x] I have left TODOs throughout the codebase, if applicable

## Summary by CodeRabbit

- **New Features**
  - Introduced a mapping for service activation heights in the Supplier structure, allowing tracking of service activation timings.
  - Added new configuration parameter for controlling resource allocation in relay services.
  - New function to retrieve the next session start height for improved session management.

- **Bug Fixes**
  - Modified logic in supplier activity checks to enhance validation by incorporating service-specific conditions.

- **Tests**
  - Refactored test suite for supplier management with new helper functions and additional test cases to cover new scenarios. 

- **Chores**
  - Enhanced mock generation directives to include session management interfaces for improved testing capabilities.

### CODE REVIEW

File: x/supplier/types/expected_keepers.go
Comment (Olshansk) [inline x/supplier/types/expected_keepers.go:38]: Need to add this to mockgen on line 1

File: proto/poktroll/shared/supplier.proto
Comment (Olshansk) [inline proto/poktroll/shared/supplier.proto:21]: Should they not be session start height?

File: x/supplier/keeper/msg_server_stake_supplier_test.go
Comment (Olshansk) [inline x/supplier/keeper/msg_server_stake_supplier_test.go:231]:
```suggestion
// stakeSupplierForServiceMsg prepares and returns a MsgStakeSupplier that stakes the given supplier
// address, stake amount, and service ID.
func stakeSupplierForServiceMsg(
```

File: x/supplier/keeper/msg_server_stake_supplier_test.go
Comment (Olshansk) [inline x/supplier/keeper/msg_server_stake_supplier_test.go:252]: Please use app/volatile/constants.go for `upokt`

File: e2e/tests/stake_supplier.feature
Comment (Olshansk) [inline e2e/tests/stake_supplier.feature:33]:
```suggestion
        And the user verifies the "supplier" for account "supplier2" is not staked
```

File: e2e/tests/stake_supplier.feature
Comment (Olshansk) [inline e2e/tests/stake_supplier.feature:34]:
```suggestion
        Then the user stakes a "supplier" with "1000070" uPOKT for "anvil" service from the account "supplier2"
```

File: e2e/tests/stake_supplier.feature
Comment (Olshansk) [inline e2e/tests/stake_supplier.feature:3]:
```suggestion
    Scenario: User can stake and unstake a Supplier waiting for it to unbond
```

File: e2e/tests/stake_supplier.feature
Comment (Olshansk) [inline e2e/tests/stake_supplier.feature:28]:
```suggestion
    Scenario: User can restake a Supplier waiting for it to become active again
```

File: x/session/keeper/query_get_session.go
Comment (Olshansk) [inline x/session/keeper/query_get_session.go:38]:
```suggestion
	// If the request specifies a block height, use it. Otherwise, use the current
	// block height.
	// Requesting a session with a block height of 0 allows to get the current session,
	// which is useful for querying from CLI.
	blockHeight := sdk.UnwrapSDKContext(ctx).BlockHeight()
	if req.BlockHeight > 0 {
		blockHeight = req.BlockHeight
	}
```

File: x/shared/types/supplier.go
Comment (Olshansk) [inline x/shared/types/supplier.go:17]:
```suggestion
// the one during which the supplier staked for that service.
```

File: x/shared/types/supplier.go
Comment (Olshansk) [inline x/shared/types/supplier.go:21]:
```suggestion
	// Service that has been staked for is not active yet.
	if s.ServicesActivationHeight[serviceId] > queryHeight {
```

**Review by Olshansk — CHANGES_REQUESTED** [review]
Not a complete review (yet) but posting partial comments so far.

File: proto/poktroll/shared/supplier.proto
Comment (red-0ne) [reply proto/poktroll/shared/supplier.proto:21]: They are! Good catch!

File: x/supplier/keeper/msg_server_stake_supplier.go
Comment (Olshansk) [inline x/supplier/keeper/msg_server_stake_supplier.go:172]: OPTIONAL NIT: Wdyt of beginning to have functions like this be receiver functions of `sharedtypes.Params` so they're easier to reuse?

**Review by Olshansk — CHANGES_REQUESTED** [review]
One small NIT. Will approve after we merge the base PR in just to sanity check the changes.

File: x/supplier/keeper/msg_server_stake_supplier.go
Comment (red-0ne) [reply x/supplier/keeper/msg_server_stake_supplier.go:172]: I'd rather have it at `x/shared/session.go` and pass in the query height.

The current `getNextSessionStartHeight` uses `sdkCtx` to get the current height which makes it non-portable to non `on-chain use cases`.

`GetNextSessionStartHeight(sharedParams *sharedtypes.Params, queryHeight int64) int64`

File: x/shared/session.go
Comment (Olshansk) [inline x/shared/session.go:159]: Great comment!

**Review by Olshansk — APPROVED** [review]

**Pull request merged**
